app/components/layout/Layout.tsx

```
@@ -39,8 +39,8 @@ export default class Layout extends Component<LayoutProps, LayoutState> {
         const { doLogout, initialPageLoad } = this.props;
         initialPageLoad()
             .then(() => {
-                StatusPoller.getPoller().start();
-                UserAppDataAutoSaver.getAutoSaver().start();
+                StatusPoller.getPoller()?.start();
+                UserAppDataAutoSaver.getAutoSaver()?.start();
```

Why do we need this change?

---

Hm... looks like the refactoring would be not straightforward. For now I would go with:
```
    static getPoller() {
        return singleton!;
    }
```
I'd prefer to see a runtime exception if `singleton === null` rather than silently passing by.

---

Can I resolve this review comment?

---

Can you somehow reply to my suggestion?

---

I tried to look how it was solved in other places in the code.
But I see that it used the same solution as I use `?.`
In Line 38 in `app/actions/userApp.ts`:
```
        const autoSaver = UserAppDataAutoSaver.getAutoSaver();
        autoSaver?.stop();
```

It was added on last Friday in the #2291 PR.

I don't see other solution apart of leaving it as it is or changing our implementation of singleton pattern.

---

Oh now I understand what you mean. Checking for null in `app/components/layout/Layout.tsx` before calling `StatusPoller.getPoller().start();`, like:
```
const poller = StatusPoller.getPoller();
if (poller === null) { throw expecption } else { StatusPoller.getPoller().start(); }
```

---

So instead of:
```typescript
StatusPoller.getPoller()?.start();
UserAppDataAutoSaver.getAutoSaver()?.start();
```
Change to:
```typescript
const statusPoller = StatusPoller.getPoller();
if (statusPoller === null) {
    throw new Error('StatusPoller.getPoller() returns null');
} else {
    statusPoller.start();
}
const userAppDataAutoSaver = UserAppDataAutoSaver.getAutoSaver();
if (userAppDataAutoSaver === null) {
    throw new Error('StatusPoller.getPoller() returns null');
} else {
    userAppDataAutoSaver.start();
}
```

On L67 and L68 I have:
```typescript
StatusPoller.getPoller()?.stop();
UserAppDataAutoSaver.getAutoSaver()?.stop();
```

So I'll need change there as well.

I tried to create a function:
```typescript
function throwErrorOnNull<T>(obj: T, methodName: string, error: string) {
    if (obj === null) {
        throw new Error(error);
    } else {
        obj[methodName]();
    }
}
```
But it didn't work.

---

Changed: [1ba53c6](https://github.com/cloudify-cosmo/cloudify-stage/pull/2247/commits/1ba53c6d34aab61615b071d96d18f1e8a3d256dc)